lib/Intlc/Linter.hs

```diff
@@ -0,0 +1,37 @@
+module Intlc.Linter where
+
+import           Intlc.ICU
+import           Prelude   hiding (Type)
+
+data Status
+  = Success
+  | Failure Text
+  deriving (Eq, Show)
+
+countInterpolations :: NEStream -> Int
+countInterpolations = foldr go 0
```

Comment:
This function will keep traversing the list and counting even if we've already counted two interpolations. We could short-circuit if we forego `foldr` and directly recurse.

Reply:
It took me longer that I care to admit but at the end I came up with b6238b77387705b2085ff66ff87e75a4b836ca5f — I'm pretty sure you'll find ways to improve this but the tests are still passing so I believe the output is right.

Reply:
Awesome! It does take some time to adjust for sure.

There is a way to improve upon this by pattern matching on the head and tail of the input list, removing the need for `i`. (It might be a tad simpler if you convert the `NEStream` to a `Stream` with `toList`.) This is well worth doing as a learning exercise as it's at the heart of iteration via recursion in Haskell.

You can also fully encapsulate the internal state:

```hs
hasManyInterpolations :: NEStream -> Bool
hasManyInterpolations = go 0 where
  go n (x:ys) = etc
```

Reply:
Thanks for the pointers. I've got https://github.com/unsplash/intlc/commit/95a27368fdf1772b5aa6b8a94b6bf349fa30a73d although there is a bit of duplication 🤔

Reply:
That's a hard one. I think most often you'd see a `null` (empty list) check with unsafe `head`/`tail` calls.

We can cheekily avoid that by moving our guard up a layer:

```hs
go n
  | n > 1     = const $ Failure TooManyInterpolations
  | otherwise = \case
    []     -> Success
    (x:xs) -> go (count x + n) xs
```

Or more simply we can pattern match on `n` directly given we know we're always incrementing by one:

```hs
go 2 _      = Failure TooManyInterpolations
go _ []     = Success
go n (x:xs) = go (count x + n) xs
```

I think the latter is quite elegant. :slightly_smiling_face:

Reply:
Oh yeah very cool! Forgot we could pattern match on ints